metrics-facade-base/src/test/java/com/ringcentral/platform/metrics/instrument/executors/MonitoredScheduledExecutorServiceTest.java

```diff
@@ -0,0 +1,379 @@
+        // when
+        timeNanosProvider.increaseSec(1);
+
+        // then
+        verify(executionTimer).stopwatch(LABEL_VALUES);
+        verify(runningCounter).inc(LABEL_VALUES);
+        verify(runningCounter).dec(LABEL_VALUES);
+        verify(completedRate).mark(LABEL_VALUES);
+        verify(executionStopwatch).stop();
+        verify(periodPercentHistogram).update(100L, LABEL_VALUES);
+        // verify(scheduledOverrunRate).mark(LABEL_VALUES);
+        verifyNoMoreMetricInteractions();
+    }
+
+    @Test
+    public void scheduleAtFixedRate_Overrun() {
+        // given
+        verifyNoMoreMetricInteractions();
+        when(executionStopwatch.stop()).thenReturn(SECONDS.toNanos(15L));
+        Runnable task = mock(Runnable.class);
+
+        // when
+        monitoredExecutor.scheduleAtFixedRate(task, 10, 10, SECONDS);
+
+        // then
+        verifyNoMoreMetricInteractions();
+
+        // when
+        timeNanosProvider.increaseSec(9);
+
+        // then
+        verifyNoMoreMetricInteractions();
+
+        // when
+        timeNanosProvider.increaseSec(1);
+
+        // then
+        verify(executionTimer).stopwatch(LABEL_VALUES);
+        verify(runningCounter).inc(LABEL_VALUES);
+        verify(runningCounter).dec(LABEL_VALUES);
+        verify(completedRate).mark(LABEL_VALUES);
+        verify(executionStopwatch).stop();
+        verify(periodPercentHistogram).update(150L, LABEL_VALUES);
+        verify(scheduledOverrunRate).mark(LABEL_VALUES);
+
+        verifyNoMoreMetricInteractions();
+    }
+
+    @Test
+    public void invokeAll() throws Exception {
+        // given
+        verifyNoMoreMetricInteractions();
+
+        Callable<Integer> task_1 = mock(Callable.class);
+        when(task_1.call()).thenReturn(1);
+
+        Callable<Integer> task_2 = mock(Callable.class);
+        when(task_2.call()).thenReturn(2);
+
+        // when
+        assertThat(monitoredExecutor.invokeAll(List.of(task_1, task_2)).stream().map(MonitoredScheduledExecutorServiceTest::value).collect(toList()), is(List.of(1, 2)));
+
+        // then
+        verifyNoMoreMetricInteractions();
+    }
+
+    private void verifyNoMoreMetricInteractions() {
+        verifyNoMoreInteractions(submittedRate, runningCounter, completedRate, idleTimer, idleStopwatch, executionTimer, executionStopwatch);
+        verifyNoMoreInteractions(scheduledOnceRate, scheduledRepetitivelyRate, scheduledOverrunRate, periodPercentHistogram);
+    }
+
+    private static <T> CompletableFuture<T> completedFutureFor(Callable<T> task) {
+        try {
+            return completedFuture(task.call());
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static Integer value(Future<Integer> future) {
+        try {
+            return future.get();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    class ScheduledExecutorServiceImpl extends TestScheduledExecutorService {
```

Comment:
We could use here spy (to be more specific in before method).
Then we don't need neither ScheduledExecutorServiceAdapter, not TestScheduledExecutorService at all.
```
ScheduledExecutorService parent = spy(ScheduledExecutorService.class);
when(parent.schedule(any(Runnable.class), anyLong(), any(TimeUnit.class)))
        .then(inv -> {
            verify(scheduledOnceRate).mark(LABEL_VALUES);
            verify(idleTimer).stopwatch(LABEL_VALUES);
            verifyNoMoreMetricInteractions();

            Runnable command = inv.getArgument(0);
            long delay = inv.getArgument(1);
            TimeUnit timeUnit = inv.getArgument(2);

            long delayNanos = timeUnit.toNanos(delay);
            TestRunnableScheduledFuture future = new TestRunnableScheduledFuture();

            if (delayNanos > 0L) {
                timeNanosProvider.addListener(t -> {
                    command.run();
                    future.run();
                }, delayNanos);
            } else {
                command.run();
                future.run();
            }

            return future;
        });

when(parent.scheduleAtFixedRate(any(Runnable.class), anyLong(), anyLong(), any(TimeUnit.class)))
        .then(inv -> {
                    verify(scheduledRepetitivelyRate).mark(LABEL_VALUES);
                    verifyNoMoreMetricInteractions();

                    Runnable command = inv.getArgument(0);
                    long initialDelay = inv.getArgument(1);
                    long delay = inv.getArgument(2);
                    TimeUnit timeUnit = inv.getArgument(3);

                    TestTimeNanosProvider.PeriodicListener listener = new TestTimeNanosProvider.PeriodicListener() {

                        @Override
                        public long periodNanos() {
                            return timeUnit.toNanos(delay);
                        }

                        @Override
                        public void whenTimeNanos(long timeNanos) {
                            command.run();
                        }
                    };

                    timeNanosProvider.addListener(listener, timeUnit.toNanos(initialDelay));
                    return new TestRunnableScheduledFuture();
                }
        );

doAnswer(inv -> {
            Runnable task = inv.getArgument(0, Runnable.class);
            verify(submittedRate).mark(LABEL_VALUES);
            verify(idleTimer).stopwatch(LABEL_VALUES);
            verifyNoMoreMetricInteractions();

            task.run();

            verify(idleStopwatch).stop();
            verify(executionTimer).stopwatch(LABEL_VALUES);
            verify(runningCounter).inc(LABEL_VALUES);
            verify(runningCounter).dec(LABEL_VALUES);
            verify(completedRate).mark(LABEL_VALUES);
            verify(executionStopwatch).stop();
            return null;
        }

).when(parent).execute(any());

doAnswer(inv -> {
            Callable<?> task = inv.getArgument(0, Callable.class);
            verify(submittedRate).mark(LABEL_VALUES);
            verify(idleTimer).stopwatch(LABEL_VALUES);
            verifyNoMoreMetricInteractions();

            Future<?> future = completedFutureFor(task);

            verify(idleStopwatch).stop();
            verify(executionTimer).stopwatch(LABEL_VALUES);
            verify(runningCounter).inc(LABEL_VALUES);
            verify(runningCounter).dec(LABEL_VALUES);
            verify(completedRate).mark(LABEL_VALUES);
            verify(executionStopwatch).stop();
            return future;
        }
).when(parent).submit(any(Callable.class));

doAnswer(inv -> {
            Collection<? extends Callable<?>> tasks = inv.getArgument(0, Collection.class);
            verify(submittedRate).mark(2L, LABEL_VALUES);
            verify(idleTimer, times(2)).stopwatch(LABEL_VALUES);
            verifyNoMoreMetricInteractions();

            List<Future<?>> collect = tasks.stream().map(MonitoredScheduledExecutorServiceTest::completedFutureFor).collect(toList());

            verify(idleStopwatch, times(2)).stop();
            verify(executionTimer, times(2)).stopwatch(LABEL_VALUES);
            verify(runningCounter, times(2)).inc(LABEL_VALUES);
            verify(runningCounter, times(2)).dec(LABEL_VALUES);
            verify(completedRate, times(2)).mark(LABEL_VALUES);
            verify(executionStopwatch, times(2)).stop();

            return collect;
        }
).when(parent).invokeAll(any(Collection.class));
```

Reply:
The problem is that you're duplicating the complex scheduling logic, the primary reason for which this `TestScheduledExecutorService` is used in several places.